### ISSUE TITLE
Deadlocks in nested transactions attempt rollback

### ISSUE THREAD

**Issue by jamiemccarthy** [rails/rails#30752]
### Steps to reproduce

If a MySQL deadlock occurs (thus raising an `ActiveRecord::Deadlocked` exception), inside a
nested transaction (i.e. a `SavepointTransaction`), Rails' rescue handler attempts to
roll back to the most recent savepoint (calling [rollback_transaction](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L233), [rollback](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L221), and then [exec_rollback_to_savepoint](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/savepoints.rb#L15)).

However, after a deadlock, the MySQL [server has already rolled back](https://dev.mysql.com/doc/refman/5.7/en/innodb-deadlock-detection.html) the entire transaction, including any savepoints:

> InnoDB automatically detects transaction deadlocks and rolls back a transaction or transactions to break the deadlock.

This means when the "ROLLBACK TO SAVEPOINT" query is sent, the savepoint is never valid, and always raises an unnecessary and incorrect [Mysql2::Error](https://github.com/brianmario/mysql2/blob/58d0a0d4cf59/lib/mysql2/error.rb#L4) with the message "SAVEPOINT active_record_n does not exist: ROLLBACK TO SAVEPOINT active_record_n".

A test file appears below.

### Expected behavior
When a MySQL deadlock happens inside a nested `within_new_transaction(:requires_new => true)` block, `within_new_transaction` should cleanly roll back and raise an `ActiveRecord::Deadlocked` exception.

### Actual behavior
Nested `within_new_transaction(:requires_new => true)` raises a `Mysql2::Error` exception.

### System configuration
**Rails version**: seen in production 5.0.2 (with slightly different symptoms), local testing on master (c9527652c753).

**Ruby version**: production 2.3.3/2.3.5, local testing 2.4.2

### Proposed fix

From my understanding of the Rails `Transaction` and `TransactionState` classes, I think the best fix would be to introduce a new [`TransactionState`](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L5) (maybe `:invalidated`?). This state would be applied to each item in the `TransactionManager`'s stack as soon as `ActiveRecord::Deadlocked` is [rescued](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L231).

When a transaction's state is `:invalidated`, each `Transaction` subclass would change the behavior of `#rollback` in the minimal possible way ([here](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L147) and [here](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L170)), by omitting the `connection.rollback_*` call but continuing with all other behavior including rolling back records (right?).

Additionally, they would change the behavior of `#commit` to raise a (different) error (a commit during an exception handler would be ill-advised and probably already raise an exception, but we can at least make the language of that exception more precise).

This same error probably also happens with Postgres. My guess is that `SerializationFailure` should be handled this way exactly as `Deadlocked` is. I haven't tried it.

### Test case

This is a simple modification of the existing [`Mysql2TransactionTest`](https://github.com/rails/rails/blob/c9527652c753/activerecord/test/cases/adapters/mysql2/transaction_test.rb).

```
ubuntu@rails-dev-box:/vagrant/rails$ cat activerecord/test/cases/adapters/mysql2/nested_deadlock_test.rb
# frozen_string_literal: true

require "cases/helper"
require "support/connection_helper"

module ActiveRecord
  class Mysql2NestedDeadlockTest < ActiveRecord::Mysql2TestCase
    self.use_transactional_tests = false

    class Sample < ActiveRecord::Base
      self.table_name = "samples"
    end

    setup do
      @abort, Thread.abort_on_exception = Thread.abort_on_exception, false

      @connection = ActiveRecord::Base.connection
      @connection.clear_cache!

      @connection.transaction do
        @connection.drop_table "samples", if_exists: true
        @connection.create_table("samples") do |t|
          t.integer "value"
        end
      end

      Sample.reset_column_information
    end

    teardown do
      @connection.drop_table "samples", if_exists: true

      Thread.abort_on_exception = @abort
    end

    test "deadlock correctly raises Deadlocked inside nested SavepointTransaction" do
      assert_raises(ActiveRecord::Deadlocked) do
        barrier = Concurrent::CyclicBarrier.new(2)

        s1 = Sample.create value: 1
        s2 = Sample.create value: 2

        begin
          thread = Thread.new do
            # Start a RealTransaction
            Sample.transaction(:requires_new => false) do
              # Start a SavepointTransaction inside it
              Sample.transaction(:requires_new => true) do
                s1.lock!
                barrier.wait
                s2.update_attributes value: 1
              end
            end
          end

          begin
            # Start a RealTransaction
            Sample.transaction(:requires_new => false) do
              # Start a SavepointTransaction inside it
              Sample.transaction(:requires_new => true) do
                s2.lock!
                barrier.wait
                s1.update_attributes value: 2
              end
            end
          ensure
            thread.join
          end

        rescue ActiveRecord::StatementInvalid => e
          if /SAVEPOINT active_record_. does not exist: ROLLBACK TO SAVEPOINT/ =~ e.to_s
            assert nil, "ROLLBACK TO SAVEPOINT query issued for savepoint that no longer exists due to deadlock: #{e}"
          else
            raise e
          end
        end
      end
    end

  end
end
```

**Comment by joevandyk** [comment]
I can reproduce this error on 5.2.

**Comment by jamiemccarthy** [comment]
For what it's worth, my company has been running a homebrew version of #30922 in production since Oct. 2017, first on 5.0 and now on 5.1, on an application that produces a fair number of deadlocks inside nested transactions.

We'll have to modify our patch to account for 5.2's [locking around transactions](https://github.com/rails/rails/commit/e4c197c7698e#diff-52c8dd8e01c039c37b33b3fcbdfe406aR182). Also, 5.2 deprecates (and 6.0 removes) the internal way of setting a TransactionState, [`#set_state`](https://github.com/rails/rails/blob/v5.2.4.1/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L43), so we'll have to figure out a clean way of doing that. But we plan to keep running our patch.

I'd be open to revisiting and resubmitting #30922 if there's any interest.

**Comment by jeremy** [comment]
Def an issue to keep fresh (and fix).

**Comment by rafaelfranca** [comment]
Fixed by 7eeab35
